File: internal/algorithm/tokenbucket/tokenbucket.lua
-- Token Bucket rate limiter
--
-- State stored as a Redis hash: { tokens (float), last_ms (integer) }
-- On each request: refill based on elapsed time, then deduct cost.
--
-- KEYS[1]  bucket key  e.g. rl:{tenant:key}tb
--
-- ARGV[1]  now_ms       current unix timestamp in milliseconds
-- ARGV[2]  capacity     maximum tokens the bucket can hold (integer)
-- ARGV[3]  refill_rate  tokens added per second (float)
-- ARGV[4]  cost         tokens to consume for this request (integer >= 1)
--
-- Returns: Redis array { allowed, remaining, retry_after_ms }
--   allowed:        1 if permitted, 0 if denied
--   remaining:      tokens remaining after this request (floor), >= 0
--   retry_after_ms: milliseconds until next token available; 0 when allowed

local now_ms      = tonumber(ARGV[1])
local capacity    = tonumber(ARGV[2])
local refill_rate = tonumber(ARGV[3])
local cost        = tonumber(ARGV[4])

local data    = redis.call('HMGET', KEYS[1], 'tokens', 'last_ms')
local tokens  = tonumber(data[1])
local last_ms = tonumber(data[2])

-- First request: start with a full bucket.
if tokens == nil then
    tokens  = capacity
    last_ms = now_ms
end

-- Refill: add tokens proportional to elapsed time, capped at capacity.
local elapsed_sec = math.max(0, (now_ms - last_ms) / 1000.0)
local refilled    = math.min(capacity, tokens + elapsed_sec * refill_rate)

if refilled < cost then
    -- Not enough tokens. Compute exact wait time.
    local deficit        = cost - refilled
    local retry_after_ms = math.ceil(deficit / refill_rate * 1000)
    -- Update stored tokens and timestamp even on deny, so the next
    -- refill calculation uses the current time as the baseline.
    redis.call('HMSET', KEYS[1], 'tokens', refilled, 'last_ms', now_ms)
    redis.call('PEXPIRE', KEYS[1], math.ceil(capacity / refill_rate * 1000) + 5000)
    return {0, math.floor(refilled), retry_after_ms}
end

local new_tokens = refilled - cost
redis.call('HMSET', KEYS[1], 'tokens', new_tokens, 'last_ms', now_ms)
-- TTL: time to fully refill from empty, plus a 5s buffer.
redis.call('PEXPIRE', KEYS[1], math.ceil(capacity / refill_rate * 1000) + 5000)

return {1, math.floor(new_tokens), 0}

File: internal/algorithm/tokenbucket/tokenbucket.go
// Package tokenbucket implements rate limiting using the token bucket algorithm.
// It allows bursting up to Capacity tokens, then refills at RefillRate tokens/second.
package tokenbucket

import (
	_ "embed"
	"context"
	"fmt"
	"time"

	"github.com/varunsalian/ratelimiter/internal/algorithm"
	"github.com/varunsalian/ratelimiter/internal/store"
)

//go:embed tokenbucket.lua
var bucketScript string

// Bucket implements the token bucket rate limiting algorithm.
//
// State in Redis: a single hash key per rate limit key holding two fields:
//   - tokens:  current token count (float — fractional tokens accumulate between requests)
//   - last_ms: unix millisecond timestamp of the last check
//
// On each request, elapsed time since last_ms is computed and tokens are refilled
// (up to Capacity). If enough tokens exist, cost is deducted and the request is allowed.
//
// Unlike sliding window algorithms, token bucket smooths bursty traffic: a client
// that was idle can burst up to Capacity requests instantly, then is rate-limited
// at RefillRate. Use SlidingWindowLog when hard-capping bursts is required.
type Bucket struct {
	script store.Scripter
}

// New creates a Bucket bound to the given store.
// Safe for concurrent use.
func New(s store.Store) *Bucket {
	return &Bucket{script: s.Script(bucketScript)}
}

// Check atomically checks and (if allowed) deducts tokens from the bucket.
//
// cfg.Capacity must be > 0. cfg.RefillRate must be > 0.
// The limit parameter is unused by this algorithm (capacity serves as the limit).
func (b *Bucket) Check(ctx context.Context, key string, _ int64, cost int32, cfg algorithm.Config) (algorithm.Result, error) {
	if cost <= 0 {
		cost = 1
	}

	nowMs := time.Now().UnixMilli()
	bucketKey := fmt.Sprintf("%stb", key)

	raw, err := b.script.Run(ctx, []string{bucketKey},
		nowMs,            // ARGV[1]: current time in milliseconds
		cfg.Capacity,     // ARGV[2]: max tokens (burst ceiling)
		cfg.RefillRate,   // ARGV[3]: tokens per second
		int64(cost),      // ARGV[4]: tokens to consume
	)
	if err != nil {
		return algorithm.Result{}, fmt.Errorf("tokenbucket: %w", err)
	}

	return parseResult(raw, cfg)
}

// parseResult converts the Lua return value {allowed, remaining, retry_after_ms} into a Result.
func parseResult(raw any, cfg algorithm.Config) (algorithm.Result, error) {
	vals, ok := raw.([]any)
	if !ok || len(vals) < 3 {
		return algorithm.Result{}, fmt.Errorf("tokenbucket: unexpected script response: got %T (len=%d)", raw, lenOf(raw))
	}

	allowed      := toInt64(vals[0]) == 1
	remaining    := toInt64(vals[1])
	retryAfterMs := toInt64(vals[2])

	// ResetAt: time until bucket is fully refilled from current remaining tokens.
	// This is an approximation — the exact value depends on live token state.
	var resetAt time.Time
	if cfg.RefillRate > 0 && cfg.Capacity > 0 {
		refillDuration := time.Duration(float64(cfg.Capacity)/cfg.RefillRate*1000) * time.Millisecond
		resetAt = time.Now().Add(refillDuration)
	}

	return algorithm.Result{
		Allowed:      allowed,
		Remaining:    remaining,
		RetryAfterMs: retryAfterMs,
		ResetAt:      resetAt,
	}, nil
}

func toInt64(v any) int64 {
	switch x := v.(type) {
	case int64:
		return x
	case int:
		return int64(x)
	default:
		return 0
	}
}

func lenOf(v any) int {
	if s, ok := v.([]any); ok {
		return len(s)
	}
	return -1
}
